guardian/support-admin-console
# Two-tier geographical targeting for amounts tests

## What does this change?
Marketing have asked us to supply them with the ability to set up a contribution amounts test for a given country, in addition to the tests for existing regions.

+ Note that this request represents a major refactor of the data model used to store amounts tests, which previously restricted amounts test data to just our 7 named regions.

Associated PRs in other repos
+ SAC: https://github.com/guardian/support-admin-console/pull/439
+ SDC: https://github.com/guardian/support-dotcom-components/pull/940
+ Support Frontend: https://github.com/guardian/support-frontend/pull/5027

This PR updates the RRCP `Amounts` page to allow Marketing colleagues to create, update and delete `country` tests in addition to existing `region tests`. It also fixes some existing UX issues with the existing page.

We also remove the RRCP `Contribution types` page as that data will now be recorded as part of the amounts test data model

## Navigating through the PR
This PR touches 45 files, as follows:

**/app and /conf folders** (7 files)
The changes in these Scala files handle typing, encoding and decoding the new amounts-v2 model (stored in S3). It also removes code handling the ContributionTypes model which has now been folded into the amounts-v2 model

**/public/src/components/amounts folder** (24 files)
I need to apologise here, as the number of files changed is a result of me renaming files to make their purpose more obvious. Sadly GitHub has failed to pickup that some new/deleted files were in fact an attempt to rename the file.

Nevertheless the bulk of the PR code resides here, as I adapted the RRCP amounts page to accommodate the management of 2-tier amounts tests - see screenshots below for more details.

**Other /public/src/components folders** (13 files)
Changes in these files are mainly minor edits for bugs arising from the move from a single amounts-test object model to an array of amounts-test objects model, and the removal of the RRCP `Contribution types` page.

**/public/src/utils folder** (1 file)
We define the JS types for the new amounts-v2 model here

## How to test
Testing is in 3 parts:
+ Functionality testing of updated RRCP amounts page on local machine, and in CODE
+ End-to-end testing in CODE to make sure that a country-specific amounts test displays correctly in Epics and Banners for a user in a country with an associated amounts test, that details of the correct amounts test get coded into the support site URL, and that clicking on the CTA will take the user to the support site which displays the same amounts data and choice as the user saw/made in the Epic/Banner. [Link to testing document](https://docs.google.com/document/d/1P-MS1Do_5Xqro6E6KKiw_7gID8c4U3L_CE_nAn-j2VI/edit).

## Screenshots

#### Amounts page before changes
+ The name of the test (used for tracking) is not clear
+ The delete buttons - both for the live test, and for each variant in the test - are unclear; clicking on the delete buttons gives no warnings that the action is permanent and irrevocable
+ No visual indication on whether a test is running in a given region, except by clicking on the test and checking its `isLive` selector
+ A live test name can only be set once
+ The default contribution type, and parameters to determine which contributions types to display on the checkout page, are set elsewhere in RRCP
+ Only a single save button for all tests, with no feedback that the save has actioned.

![Screenshot 2023-05-10 at 16 01 39](https://github.com/guardian/support-admin-console/assets/5357530/efaa01af-8cb0-4732-b4d7-64c1efd5d5c0)

#### New Amounts page - initial load
+ The page no longer loads with one of the amounts tests pre-loaded
+ All amounts tests are displayed in the left hand column, with region tests shown above country tests
+ every test displays its `isLive` status by colour (live tests have red borders)
+ new `create test` button at top of the country tests list

![Screenshot 2023-05-10 at 16 02 18](https://github.com/guardian/support-admin-console/assets/5357530/8608ad7b-2f8c-4768-ba18-968c41dfa3d8)

+ Country tests are created using an interactive selector which lists all countries for which no test currently exists

![Screenshot 2023-05-10 at 16 48 40](https://github.com/guardian/support-admin-console/assets/5357530/5b6721fe-71bf-4509-b25b-d3651de93df8)

**FOR ALL TESTS**
+ The test tracking names are displayed at the top of the form; the Live A/B test name is editable
+ The `save test` and `add variant` buttons appear twice, near the top of the form, and at the bottom
+ The `display contribution type` checkboxes are displayed at the top of each variant.
+ The `default contribution type` buttons selector is displayed at the top of each variant.
+ Each variant panel, except the `control` variant, has a clearly marked `delete variant` button at the top of the panel
+ Clicking the `delete variant` button triggers a confirm action modal to display
+ The `save test` button will be enabled after the first change is made to the test's data

![Screenshot 2023-05-15 at 15 30 30](https://github.com/guardian/support-admin-console/assets/5357530/588a38f9-91e9-469e-ab1f-12a58d3db008)

**ADDITIONALLY FOR COUNTRY-SPECIFIC TESTS**
+ The test's names are defined automatically when the test is first created; the Live test name is editable
+ A `delete test` button appears at the top and bottom of the test's form; clicking the button will launch a warning modal asking for the action to be confirmed

![Screenshot 2023-05-15 at 15 30 56](https://github.com/guardian/support-admin-console/assets/5357530/b54a74b5-5915-4b7a-96ee-d3e8d0d22569)
